Add dropdown on signup and login to move between servers

nospame opened the pull request:
## Product Description
On login page: 
<img width="494" height="481" alt="image" src="https://github.com/user-attachments/assets/acbfda61-3fb3-4f2b-bd92-39b575f569c8" />

On signup page:
<img width="491" height="480" alt="image" src="https://github.com/user-attachments/assets/17c6c02e-9242-43de-80ee-787bd2c5bad3" />
<img width="493" height="476" alt="image" src="https://github.com/user-attachments/assets/7fb73917-f944-45fc-ab01-0761275a8291" />



## Technical Summary
https://dimagi.atlassian.net/browse/SAAS-17739, https://dimagi.atlassian.net/browse/SAAS-18052
Conditionally adds a new form field `server_location` to the new user registration form and the hq login form -- but not the domain-specific version of the login form. Uses knockout to bind to the value of this form field and, when the value changes, swap out the subdomain component of the current url in the user's browser with the new value.

Adds a bunch of styling to make the new select2 in the login form look more like the custom `RadioSelect` crispyform widget we use in the registration form.

## Feature Flag
Not feature flagged. Considered adding a `FeatureRelease` flag but we can simply hold on deploying until the desired release date.

## Safety Assurance

### Safety story
Despite a good number of lines of code added, the functionality here is not crazy. Add a form field, get the value of the field, interpolate it into the current url and navigate to that url. There is some risk to modifying existing forms, especially important ones like registration and login. Automated testing for aspects of those forms helps somewhat to protect against regression.

Tested functionality locally by changing `settings.SERVER_ENVIRONMENT` to eu, production, or india or leaving it set to the local environment name and seeing that the dropdown appears or is hidden and calls JS functions as expected.  Tested functionality on staging by temporarily adding a `STAGING` option to `ServerLocation` and seeing that it appears correctly and navigates between servers in a live environment as expected. Also ran through sign up and login workflows to see they are not affected.

### Automated test coverage
Basic automated tests added for new JS component.

### QA Plan
Not planning it.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Add ServerLocation model

nospame committed:
Conditionally add server location field to registration form

nospame committed:
Add can_select_server check to HQLoginView

nospame committed:
Conditianally add server location field to login form

nospame committed:
Add server_location_select javascript component

nospame committed:
Add server location js functionality to registration page

nospame committed:
Add can_select_server to initial page data for login view

nospame committed:
Add server location js functionality to login form

Some piecing together of models to create something for koApplyBindings; it's possible for either, both, or neither of sso login and server location select to be applicable

nospame committed:
Style server location select2 on login page to look similar to signup and other form fields

nospame committed:
Prettify html and stylesheets

nospame committed:
Bootstrap 5 Migration - Rebuilt diffs

nospame committed:
Default can_select_server kwarg to False

Fixes KeyError when kwarg doesn't exist, as in existing EmailAuthenticationForm tests

nospame committed:
Only add can_select_server kwarg if step is auth

nospame committed:
Add automated tests for server_location_select.js

jingcheng16 commented on corehq/apps/registration/static/registration/js/login.js:
```diff
@@ -22,17 +24,37 @@ $(function () {
+    var enforceSsoLogin = initialPageData.get('enforce_sso_login');
+    var canSelectServer = initialPageData.get('can_select_server');
+
+    var loginFormModel = {};
```
`var` to `let`

jingcheng16 reviewed (approved):
no blocking feedbacks.

nospame committed:
Use const and let in login.js

nospame committed:
ServerLocation: don't inherit from object

nospame committed:
setWindowHref -> navigateTo

nospame committed:
Don't show server location select when there is a 'next' url

More likely than not this is a domain-specific url, but if it's anything with any id at all it probably doesn't exist on the other servers

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into ejp/server-location-selection

nospame committed:
Use ServerLocation as source of truth for env: subdomain urls

nospame committed:
Use ServerLocation for subdomain: env urls

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into ejp/server-location-selection

nospame commented:
Closing this because this design is changing significantly, along with needing to rebase this branch, and don't feel the need to notify watchers with every change. Will re-open or make a new PR in the near future.
